### PR TITLE
exceptional exceptions

### PR DESCRIPTION
**Pull request by skshetry**
Proposing a `__pretty_exc__` data model for exceptions that can provide prettier exceptions. 

With an example on #5371, the whole thing is a bit fragile though.
Also related: https://github.com/iterative/dvc/issues/6285

---

Screen Shot 2021-09-20 at 14 30 36
Screen Shot 2021-09-20 at 14 31 17
Screen Shot 2021-09-20 at 14 37 42

### CODE REVIEW

File: dvc/logger.py [hunk 1]
```diff
@@ -142,9 +142,22 @@ def handleError(self, record):
         super().handleError(record)
         raise LoggingException(record)
 
+    def emit_pretty_exception(self, exc, verbose: bool = False):
+        return exc.__pretty_exc__(verbose=verbose)
```
Comment (skshetry) [inline dvc/logger.py:146]: A better solution would be to just handle it in `dvc.main` and introduce `DvcException.pretty_print()` kinda API, but right now, we just use `logger.exception` everywhere which makes it tricky, will have to get rid of that. For now, we are stuck with this hack.

**Review by pared — APPROVED** [review]
Nice!

**Review by efiop — APPROVED** [review]
+1, great stuff.

**Comment by skshetry** [comment]
> +1, great stuff. Btw, IIRC rich is doing that same thing? Or was it some other project?

Rich has a [`__rich_repr__`](https://rich.readthedocs.io/en/stable/pretty.html#rich-repr-protocol) protocol with which you can add support for rich output for custom objects.

This might be temporary, I am more inclined now to introduce `DvcException.show()` similar to in [`Click`](https://click.palletsprojects.com/en/8.0.x/exceptions/#which-exceptions-exist), which we can handle in `dvc.main.main`. But this requires us to get rid of `logger.exception` calls that we have.

**Comment by dberenbaum** [comment]
@skshetry Any idea why I get the old output when trying out the example in #6285?

```
$ dvc -v status
ERROR: 'dvc.yaml' format error: extra keys not allowed @ data['stages']['train']['cmd']
$ dvc doctor
DVC version: 2.7.4+cace8e
---------------------------------
Platform: Python 3.9.5 on macOS-11.5.2-arm64-arm-64bit
Supports:
        s3 (s3fs = 2021.8.0, boto3 = 1.17.5)
Cache types: reflink, hardlink, symlink
Cache directory: apfs on /dev/disk3s1s1
Caches: local
Remotes: local, s3
Workspace directory: apfs on /dev/disk3s1s1
Repo: dvc, git
```

**Comment by skshetry** [comment]
@dberenbaum, schema validation is not in this PR, only limited to the yaml syntax errors.

**Pull request merged**